File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> umap;
        for(auto i:nums){
            umap[i]++;
        }

        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;

        for(auto i:umap){
            pq.push({i.second,i.first});
            if(pq.size()>k) pq.pop();
        }
        vector<int> result;
        while(!pq.empty()){
            result.push_back(pq.top().second);
            pq.pop();
        }
     
        return result;

    }
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer,Integer> umap=new HashMap<>();

        for(int i:nums){
            umap.put(i,umap.getOrDefault(i,0)+1);
        }

        PriorityQueue<Map.Entry<Integer,Integer>> pq=new PriorityQueue<>((a,b) -> a.getValue() - b.getValue());

        for(Map.Entry<Integer,Integer> entry:umap.entrySet()){
            pq.offer(entry);
            if(pq.size()>k) pq.poll();
        }

        int result[]=new int[k];
        int index=0;
        while(!pq.isEmpty()){
            result[index++]=pq.poll().getKey();
        }
        return result;
    }
}